### PR TITLE
fix(rds): customizing secret results in unusable password and lost attachment

### LINKED ISSUE #11040: [RDS] Issue with credentials & Secret attachment
When updating a SecretsManager Secret defined in "credentials" prop of an RS Instance, we lost the attachment.

### Reproduction Steps

Create an RDS Instance with "credentials" prop targeting a Secret.
Update the Secret.

### What did you expect to happen?

I want to update my Secret but keep the attachment ith the DB.

### What actually happened?

You lost the attachment and it's not possible to add it because we have an error :
`Secret is already attached to a target.`

So a workaround is to update the DB with no/or another Secret.
But in that case, DB Instance is replaced/recreated (???!!!)

### Environment

  - **CLI Version      :** 1.68.0
  - **Framework Version:** 6.14.8
  - **Node.js Version:** v12.15.0
  - **OS               :** Linux
  - **Language (Version):** TypeScript

### Other

In my case, I've update the DB without the prop "credentials", then, update it again with my Secret (the Original one).
But it recreate @ll my DB instances 2 times and lost datas of course...

`credentials: rds.Credentials.fromSecret(SecretsManager)`

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by jogold**
When the `excludeCharacters` prop is updated the secret is correctly
regenerated but the database is not updated because the
`MasterUserPassword` prop does not change. Moreover the connection
fields created by the attachment are lost because the
`AWS::SecretsManager::SecretAttachment` doesn't "rerun" (it references
the same secret).

Introduce `fromGeneratedSecret()` and `fromPassword()` static methods
in `Credentials`. When used the `MasterUsername` prop of the database
references the username as a string instead of a dynamic reference to
the username field of the secret. Moreover the logical id of the secret
is a hash of its customization options. This way the secret gets replaced
when it's customized, the database gets updated and the attachement reruns.
This without updating the `MasterUsername` prop, avoiding a replacement
of the database.

For instances that were created from a snapshot the `MasterUsername` prop
is not specified so there's no replacement risk. Add a new static method
`fromGeneratedSecret()` in `SnapshotCredentials` to replace the secret
when its customization options change (also hash in logical id).

Deprecate `Credentials.fromUsername()` but keep it as default to avoid
a breaking change that would replace existing databases that were not
created from a snapshot.

Deprecate `SnapshotCredentials.fromGeneratedPassword()` which doesn't
replace the secret when customization options are changed.

Closes #11040

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-rds/test/test.instance.ts [hunk 1]
```diff
@@ -301,7 +301,7 @@ export = {
       expect(stack).to(haveResourceLike('AWS::RDS::DBInstance', {
         MasterUsername: ABSENT,
         MasterUserPassword: {
-          'Fn::Join': ['', ['{{resolve:secretsmanager:', { Ref: 'InstanceSecret478E0A47' }, ':SecretString:password::}}']],
+          'Fn::Join': ['', ['{{resolve:secretsmanager:', { Ref: '29ae4bfb9ee908ebe1611900a1e2469e' }, ':SecretString:password::}}']],
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/test.instance.ts:304]: Why is this change needed? Nothing changed in the body of this test, so this is very worrying that this needs updating... This would mean a database replacement for a customer when they upgrade...

File: packages/@aws-cdk/aws-rds/test/integ.instance.lit.ts [hunk 2]
```diff
@@ -49,7 +49,7 @@ class DatabaseInstanceStack extends cdk.Stack {
       instanceType: ec2.InstanceType.of(ec2.InstanceClass.BURSTABLE3, ec2.InstanceSize.MEDIUM),
       multiAz: true,
       storageType: rds.StorageType.IO1,
-      credentials: rds.Credentials.fromUsername('syscdk'),
+      credentials: rds.Credentials.fromFixedUsername('syscdk'),
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/integ.instance.lit.ts:52]: Can you add a new test instead, and leave the old one unchanged? This is such a risky change, we need all the help we can get here...

**Review by skinny85 — CHANGES_REQUESTED** [review]
Thanks for your efforts @jogold ! Some initial comments I have.

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts [hunk 3]
```diff
@@ -55,5 +66,17 @@ export class DatabaseSecret extends secretsmanager.Secret {
         excludeCharacters: props.excludeCharacters ?? DEFAULT_PASSWORD_EXCLUDE_CHARS,
       },
     });
+
+    if (props.overrideLogicalId) {
+      const hash = crypto.createHash('md5');
+      hash.update(JSON.stringify({
+        path: this.node.path, // make it unique
+        // Use here the options that influence the password generation
+        excludeCharacters: props.excludeCharacters,
```
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts:75]: We don't need to hash props that do not influence a password regeneration. If those change we can keep the same secret.

File: packages/@aws-cdk/aws-rds/lib/instance.ts [hunk 4]
```diff
@@ -1032,6 +1037,9 @@ export class DatabaseInstanceFromSnapshot extends DatabaseInstanceSource impleme
         username: credentials.username,
         encryptionKey: credentials.encryptionKey,
         excludeCharacters: credentials.excludeCharacters,
+        // Always replace secret when customization options are changed
+        // (MasterUsername is not specified for an instance created from a snapshot)
+        overrideLogicalId: true,
```
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:1042]: Yes because `MasterUsername` is not specified when creating from a snapshot so there's no replacement risk here. It will update the credentials though (both the secret and the database), is this problematic?

File: packages/@aws-cdk/aws-rds/lib/instance.ts [hunk 5]
```diff
@@ -961,7 +964,9 @@ export class DatabaseInstance extends DatabaseInstanceSource implements IDatabas
       ...this.sourceCfnProps,
       characterSetName: props.characterSetName,
       kmsKeyId: props.storageEncryptionKey && props.storageEncryptionKey.keyArn,
-      masterUsername: credentials.username,
+      masterUsername: props.credentials?.usernameAsString
+        ? props.credentials.username
+        : credentials.username,
```
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:969]: Agree, will see how to improve this.

File: packages/@aws-cdk/aws-rds/test/test.instance.ts (hunk 1 above)
Comment (jogold) [reply packages/@aws-cdk/aws-rds/test/test.instance.ts:304]: It's a `DatabaseFromSnaphot()` so its secret logical id is now a hash. As the test shows `MasterUsername` is expected to be `ABSENT` so there won't be any replacement.

File: packages/@aws-cdk/aws-rds/lib/props.ts [hunk 6]
```diff
@@ -148,10 +148,24 @@ export interface CredentialsFromUsernameOptions {
  * Username and password combination
  */
 export abstract class Credentials {
+  /**
+   * Creates Credentials for the given username, and optional password and key.
+   * If no password is provided, one will be generated and stored in SecretsManager.
+   */
+  public static fromFixedUsername(username: string, options: CredentialsFromUsernameOptions = {}): Credentials {
```
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/props.ts:155]: The default is currently handled in the instance/cluster code when `credentials` is not specified. To avoid a breaking change it still falls back to `fromUsername()`.

File: packages/@aws-cdk/aws-rds/lib/instance.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:969]: The problem I have here is that the code currently calls `fromSecret(new DatabaseSecret(...))` when the user specifies `fromUsername()` without password.

`fromSecret()` expects a `secretsmanager.Secret` so it cannot know when to render the username as a string. To do that I'll have to add an additional `username` argument to `fromSecret()` and specify it based on the new `usernameAsString`. This means that it's still the instance responsibility? Any suggestions?

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts:75]: So shouldn't we include `username` and `masterSecret`? https://github.com/aws/aws-cdk/blob/122c68c89838a9f5c4755d909635d8407f9a8019/packages/%40aws-cdk/aws-rds/lib/database-secret.ts#L48-L56

File: packages/@aws-cdk/aws-rds/lib/props.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/props.ts:155]: I wonder whether it makes sense for customers to do something like this:

```ts
new rds.DatabaseCluster(this, 'Cluster', {
  // ...
  credentials: rds.Credentials.fromFixedUsername({ excludeCharacters: '!&*^#@()' }),
});
```

"I don't care about the username - just use the default, but I want to change something about the other properties"

Does something like this make sense?

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts:75]: I actually thought about that:

`masterSecret` is only used for the multi user rotation scheme so a secret with it never represents master credentials and we can just le CF update the secret.

`username` I'm less sure but as RDS doesn't support changing the username without replacing the database, do we need to cover this? Not hashing it sorts of protect the database from a username change?

File: packages/@aws-cdk/aws-rds/lib/instance.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:969]: How about creating a new, module-private factory for `Credentials` that we can use here? Like `_fromSecretUsernameRendered(secret: ISecret, usernameAsString: boolean): Credentials`?

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts:75]: > `masterSecret` is only used for the multi user rotation scheme so a secret with it never represents master credentials and we can just let CF update the secret. There's no reason to force a replacement. Also this prop cannot be set when using the `Credentials` class.

Not sure I understand the "is only used for the multi user rotation scheme"... what about this?

```ts
    const secret = new rds.DatabaseSecret(stack, 'UserSecret', {
      username: 'user-username',
      masterSecret,
      excludeCharacters: '"@/\\',
    });

    const masterSecret = new rds.DatabaseInstance(stack, 'MasterSecret', {
      // ...
      credentials: rds.Credentials.fromSecret(secret),
    });
```

?

File: packages/@aws-cdk/aws-rds/lib/instance.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:969]: We'll (also) need to pass the `username`, right? 

Because you cannot extract a field out of a secret without creating a dynamic reference.

[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:969]: Sure 🙂

Then maybe it's enough to have a `username?: string` as the second argument, and if it's `undefined`, you use the reference to the Secret? I'm sure you'll figure out the details 🙂.

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts:75]: Yes you can actually write that code. But (1) if the user creates its own secret we are never going to control its logical id with a hash and so never control when its replaced instead of updated and (2) specifying a master secret is used to create username (vs master) credentials that use a master secret for their rotation in the multi rotation scheme. It doesn't make sense to specify a master secret for master credentials.

[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts:75]: OK. If you're confident that we only need `excludeCharacters`, I'm fine with that 🙂.

[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts:75]: If at the some point we allow to customize something like `passwordLength` then it should be added.

[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts:75]: Can you include that as a comment on the code?

File: packages/@aws-cdk/aws-rds/lib/props.ts [hunk 7]
```diff
@@ -186,6 +200,12 @@ export abstract class Credentials {
    */
   public abstract readonly username: string;
 
+  /**
+   * Whether the username should be referenced as a string and not as a dynamic
+   * reference to the username in the secret.
```
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/props.ts:205]: On a class property?

File: packages/@aws-cdk/aws-rds/lib/props.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/props.ts:155]: Since `Credentials` has a `username` property that cannot be undefined, I think this change means that we need to make `Credentials` "engine aware" to find the right default?

File: packages/@aws-cdk/aws-rds/lib/instance.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/instance.ts]: Chosen to add a optional `username` to `fromSecret` which allows users using `fromSecret` to also have the behavior of the username referenced as a string.

The same kind of logic is now a few lines above.

**Comment by jogold** [comment]
@skinny85 I finally decided to introduce two new static methods to replace `fromUsername()`: `fromGeneratedPassword()` and  `fromPassword()`. The naming is better than `fromFixedUsername()` which has no real meaning for people starting with `aws-rds`.

Since making `username` optional in `Credentials` is a breaking change those methods still require a username.

I also moved common code to an util.

File: packages/@aws-cdk/aws-rds/lib/props.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/props.ts:228]: Yep.

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 8]
```diff
@@ -489,14 +488,7 @@ export class DatabaseCluster extends DatabaseClusterNew {
...
 
-    let credentials = props.credentials ?? Credentials.fromUsername(props.engine.defaultUsername ?? 'admin');
-    if (!credentials.secret && !credentials.password) {
-      credentials = Credentials.fromSecret(new DatabaseSecret(this, 'Secret', {
-        username: credentials.username,
-        encryptionKey: credentials.encryptionKey,
-        excludeCharacters: credentials.excludeCharacters,
-      }));
-    }
+    const credentials = renderCredentials(this, props.credentials, props.engine.defaultUsername);
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:491]: Can you actually make `renderCredentials` take an `IEngine` instead? And then call `engine.defaultUsername` in the implementation (less duplication this way).

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts [hunk 9]
```diff
@@ -55,5 +66,20 @@ export class DatabaseSecret extends secretsmanager.Secret {
...
+        path: this.node.path, // make it unique
+        // Use here the options that influence the password generation.
+        // If at some point we add other password customization options
+        // they sould be added here below (e.g. `passwordLength`).
+        excludeCharacters: props.excludeCharacters,
+      }));
+      const logicalId = `${this.node.path.replace(/([^A-Za-z0-9]|\/)/g, '')}${hash.digest('hex')}`;
+
+      const secret = this.node.defaultChild as secretsmanager.CfnSecret;
+      secret.overrideLogicalId(logicalId.slice(0, 255));
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/database-secret.ts:82]: Hmm... I think you want to take the _last_ 255 characters, not the first ones, no? Otherwise, you risk losing the hash?

How about using the new `addr` method of `ConstructNode` instead? Should simplify things. Maybe with combination with `id`, to make it readable? (I'm not sure what `addr` returns, too new 😜).

File: packages/@aws-cdk/aws-rds/lib/props.ts [hunk 10]
```diff
@@ -144,14 +135,55 @@ export interface CredentialsFromUsernameOptions {
...
+    };
+  }
 
   /**
    * Creates Credentials for the given username, and optional password and key.
    * If no password is provided, one will be generated and stored in SecretsManager.
+   *
+   * @deprecated use `fromGeneratedPassword()` or `fromPassword()` for new deployments only.
+   *   Switching to `fromGeneratedPassword()` or `fromPassword()` for existing instances/clusters
+   *   will have them replaced.
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/props.ts:186]: I feel like it's important we're very precise with the wording here.

```suggestion
   * @deprecated use `fromGeneratedPassword()` or `fromPassword()` for new Clusters and Instances.
   *   Note that switching from `fromUsername()` to `fromGeneratedPassword()` or `fromPassword()` for already deployed
   *  Clusters or Instances will result in their replacement!
```

File: packages/@aws-cdk/aws-rds/lib/props.ts [hunk 11]
```diff
@@ -171,10 +203,15 @@ export abstract class Credentials {
    *   "password": <required: password>,
    * }
    * ```
+   *
+   * @param secret The secret where the credentials are stored
+   * @param username The username defined in the secret. If specified the username
+   * will be referenced as a string and not as dynamic reference to the username
+   * field in the secret
```
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/props.ts:210]: Still here 🙂

**Review by skinny85 — COMMENTED** [review]
Very nice. We're almost there. A couple of minor notes.

File: packages/@aws-cdk/aws-rds/test/test.instance.ts [hunk 12]
```diff
@@ -1138,4 +1138,64 @@ export = {
...
+        ],
+      },
+    }));
+
+    // Check that the logical id of the second secret is different
+    const db1Secret = db1.node.tryFindChild('Secret') as rds.DatabaseSecret;
+    const cfnSecret1 = db1Secret.node.defaultChild as cdk.CfnResource;
+    const db2Secret = db2.node.tryFindChild('Secret') as rds.DatabaseSecret;
+    const cfnSecret2 = db2Secret.node.defaultChild as cdk.CfnResource;
+    test.notEqual(cfnSecret1.logicalId, cfnSecret2.logicalId);
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/test.instance.ts:1180]: `logicalIds` are tokens, so you need to resolve them first before you do the comparison (try it be attempting to make this test fail by changing the hashing algorithm, it shouldn't work - this assertion should always pass).

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts (hunk 9 above)
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts:82]: `addr` returns a 42 characters hexadecimal string. Lets do `${id}${hash}` and take the last 255 chars.

File: packages/@aws-cdk/aws-rds/README.md [hunk 13]
```diff
@@ -146,13 +146,13 @@ const engine = rds.DatabaseInstanceEngine.postgres({ version: rds.PostgresEngine
 new rds.DatabaseInstance(this, 'InstanceWithUsername', {
   engine,
   vpc,
-  credentials: rds.Credentials.fromUsername('postgres'), // Creates an admin user of postgres with a generated password
+  credentials: rds.Credentials.fromGeneratedPassword('postgres'), // Creates an admin user of postgres with a generated password
 });
 
-new rds.DatabaseInstance(this, 'InstanceWithUsernameAndPassword', {
+new rds.DatabaseInstance(this, 'InstanceWithPassword', {
```
[the remark this answers is not in the document]
Comment (jogold) [reply packages/@aws-cdk/aws-rds/README.md]: reverted

File: packages/@aws-cdk/aws-rds/test/test.instance.ts (hunk 12 above)
Comment (jogold) [reply packages/@aws-cdk/aws-rds/test/test.instance.ts]: of course! moved these kind of assertions to `test.database-secret.ts` where it actually belongs.

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts [hunk 14]
```diff
@@ -55,5 +66,20 @@ export class DatabaseSecret extends secretsmanager.Secret {
...
     });
+
+    if (props.overrideLogicalId) {
+      const hash = crypto.createHash('md5');
+      hash.update(JSON.stringify({
+        path: this.node.path, // make it unique
+        // Use here the options that influence the password generation.
+        // If at some point we add other password customization options
+        // they sould be added here below (e.g. `passwordLength`).
+        excludeCharacters: props.excludeCharacters,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/database-secret.ts:77]: So `excludeCharacters` is actually optional. Which means this object will have `undefined` there as the default. Is that a mistake? Should the default value be filled here? (See line 66 above)

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts [hunk 15]
```diff
@@ -55,5 +66,20 @@ export class DatabaseSecret extends secretsmanager.Secret {
...
+      }));
+      const logicalId = `${this.node.id.replace(/[^A-Za-z0-9]/g, '')}${hash.digest('hex')}`;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/database-secret.ts:79]: I feel like `id` is too short here. You'll wind up with `Secret<some-hash>` everywhere, which is too ambiguous. Can you use `uniqueId` here instead? (No need for the `replace` part then).

**Review by skinny85 — COMMENTED** [review]
We're almost there! I have 2 minor comments/questions.

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts [hunk 16]
```diff
@@ -33,6 +34,16 @@ export interface DatabaseSecretProps {
...
+
+  /**
+   * Whether to override the logical id of the AWS::SecretsManager::Secret
+   * with a hash of the options that influence the password generation. This
+   * way a new secret will be created when the password is regenerated and the
+   * cluster or instance consuming this secret will have its credentials updated.
+   *
+   * @default false
+   */
+  readonly overrideLogicalId?: boolean;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/database-secret.ts:46]: I wonder whether a name like `replaceOnPasswordChanges` might be better here?

**Comment by jogold** [comment]
@skinny85 I think we're good here.

I added a BREAKING CHANGE to inform that the master password of instance created from snapshot will be updated. Can you check it?

@njlynch you might want to have a look at this PR since you introduced the `Credentials` class. Thanks!

**Comment by skinny85** [comment]
> I added a BREAKING CHANGE to inform that the master password of instances created from snapshot will be updated. Can you check it?

This module is now stable. We can't have breaking changes to it. Please revert the changes to `DatabaseInstanceFromSnapshot`.

**Comment by jogold** [comment]
> > I added a BREAKING CHANGE to inform that the master password of instances created from snapshot will be updated. Can you check it?
> 
> This module is now stable. We can't have breaking changes to it. Please revert the changes to `DatabaseInstanceFromSnapshot`.

Got it. But do you really consider this as a breaking change? If the secret is consumed programmatically it will have no impact... your call.

File: packages/@aws-cdk/aws-rds/README.md [hunk 17]
```diff
@@ -146,13 +146,13 @@ const engine = rds.DatabaseInstanceEngine.postgres({ version: rds.PostgresEngine
...
   vpc,
-  credentials: rds.Credentials.fromUsername('postgres'), // Creates an admin user of postgres with a generated password
+  credentials: rds.Credentials.fromGeneratedPassword('postgres'), // Creates an admin user of postgres with a generated password
 });
 
 new rds.DatabaseInstance(this, 'InstanceWithUsernameAndPassword', {
   engine,
   vpc,
-  credentials: rds.Credentials.fromUsername('postgres', { password: SecretValue.ssmSecure('/dbPassword', 1) }), // Use password from SSM
+  credentials: rds.Credentials.fromPassword('admin', SecretValue.ssmSecure('/dbPassword', '1')), // Use password from SSM
```
Comment (njlynch) [inline packages/@aws-cdk/aws-rds/README.md:155]: `admin` is not a valid username for PostgreSQL databases, so this username needs to be reverted back.

```suggestion
  credentials: rds.Credentials.fromPassword('postgres', SecretValue.ssmSecure('/dbPassword', '1')), // Use password from SSM
```

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts [hunk 18]
```diff
@@ -33,6 +34,18 @@ export interface DatabaseSecretProps {
...
+   * Whether to replace this secret when the password changes.
+   *
+   * This is achieved by overriding the logical id of the AWS::SecretsManager::Secret
+   * with a hash of the options that influence the password generation. This
+   * way a new secret will be created when the password is regenerated and the
+   * cluster or instance consuming this secret will have its credentials updated.
+   *
+   * @default false
+   */
+  readonly replaceOnPasswordChanges?: boolean;
```
Comment (njlynch) [inline packages/@aws-cdk/aws-rds/lib/database-secret.ts:48]: Should this be more appropriately named `replaceOnPasswordCriteriaChanges?` or something like that? I'm afraid the name and documentation as-is may mislead people to think the secret may be replaced when rotation is configured.

**Comment by jogold** [comment]
> > > I added a BREAKING CHANGE to inform that the master password of instances created from snapshot will be updated. Can you check it?
> > 
> > 
> > This module is now stable. We can't have breaking changes to it. Please revert the changes to `DatabaseInstanceFromSnapshot`.
> 
> Got it. But do you really consider this as a breaking change? If the secret is consumed programmatically or if rotation is in place it will have no impact... your call. The password will change but the new one is available in Secrets Manager, you never loose access.

If you want to avoid this we can expose the `replaceOnPasswordChanges` prop in `SnapshotCredentials.fromGeneratedPassword()` so that users can opt-in for the new behavior.

File: packages/@aws-cdk/aws-rds/lib/database-secret.ts (hunk 18 above)
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts:48]: I'm ok with this name, it's indeed better. @skinny85 do you validate this name?

Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/database-secret.ts:48]: Yes 🙂

**Comment by skinny85** [comment]
> > > > I added a BREAKING CHANGE to inform that the master password of instances created from snapshot will be updated. Can you check it?
> > > 
> > > 
> > > This module is now stable. We can't have breaking changes to it. Please revert the changes to `DatabaseInstanceFromSnapshot`.
> > 
> > 
> > Got it. But do you really consider this as a breaking change? If the secret is consumed programmatically or if rotation is in place it will have no impact... your call. The password will change but the new one is available in Secrets Manager, you never loose access.
> 
> If you want to avoid this we can expose the `replaceOnPasswordChanges` prop in `SnapshotCredentials.fromGeneratedPassword()` so that users can opt-in for the new behavior.

I don't love that. How about new static factory methods in `SnapshotCredentials`, similar to the new static factory methods in `Credentials`?

**Comment by jogold** [comment]
> I don't love that. How about new static factory methods in `SnapshotCredentials`, similar to the new static factory methods in `Credentials`?

Any suggestion for the name? I would expect `fromGeneratedPassword()` in `Credentials` and `SnapshotCredentials` to behave the same way but it will not be the case if we add a new static method in `SnapshotCredentials`.

How about `fromGeneratedSecret()` both in `Credentials` and `SnapshotCredentials` then?

**Comment by skinny85** [comment]
> > I don't love that. How about new static factory methods in `SnapshotCredentials`, similar to the new static factory methods in `Credentials`?
> 
> Any suggestion for the name? I would expect `fromGeneratedPassword()` in `Credentials` and `SnapshotCredentials` to behave the same way but it will not be the case if we add a new static method in `SnapshotCredentials`.
> 
> How about `fromGeneratedSecret()` both in `Credentials` and `SnapshotCredentials` then?

That's a good point about consistency between `Credentials` and `SnapshotCredentials`.

I like `fromGeneratedSecret()` well enough.

**Review by skinny85 — APPROVED** [review]
LGTM, thanks for taking care of this @jogold !

**Comment by jogold** [comment]
@skinny85 can you reapprove here? Had to merge from master because the build was stuck...

**Review by skinny85 — APPROVED** [review]
